storage/aptosdb/src/db_debugger/truncation/mod.rs

```diff
@@ -0,0 +1,311 @@
+                    self.target_version
+                ),
+            );
+
+        Self::truncate_state_merkle_db(state_merkle_db, state_merkle_target_version)?;
+
+        Self::truncate_ledger_db(
+            ledger_db,
+            ledger_db_version,
+            self.target_version,
+            self.ledger_db_batch_size,
+        )?;
+
+        if state_merkle_target_version < self.target_version {
+            // TODO(grao): If state merkle db is not consistent with ledger db, replay the write set to
+            // make them consistent.
+        }
+
+        Ok(())
+    }
+
+    fn get_current_version_in_ledger_db(ledger_db: &DB) -> Result<Option<Version>> {
+        let mut iter = ledger_db.iter::<TransactionInfoSchema>(ReadOptions::default())?;
+        iter.seek_to_last();
+        Ok(iter.next().transpose()?.map(|item| item.0))
+    }
+
+    fn get_current_version_in_state_merkle_db(state_merkle_db: &DB) -> Result<Option<Version>> {
+        Self::find_tree_root_at_or_before(state_merkle_db, u64::max_value())
+    }
+
+    fn find_tree_root_at_or_before(
+        state_merkle_db: &DB,
+        version: Version,
+    ) -> Result<Option<Version>> {
+        let mut iter = state_merkle_db.rev_iter::<JellyfishMerkleNodeSchema>(Default::default())?;
+        iter.seek_for_prev(&NodeKey::new_empty_path(version))?;
+        Ok(iter.next().transpose()?.map(|item| item.0.version()))
+    }
+
+    fn truncate_ledger_db(
+        ledger_db: DB,
+        current_version: Version,
+        target_version: Version,
+        batch_size: usize,
+    ) -> Result<()> {
+        let ledger_db = Arc::new(ledger_db);
+        let event_store = EventStore::new(Arc::clone(&ledger_db));
+        let transaction_store = TransactionStore::new(Arc::clone(&ledger_db));
+
+        let mut current_version = current_version;
+        while current_version > target_version {
+            let start_version =
+                std::cmp::max(current_version - batch_size as u64 + 1, target_version + 1);
+            let end_version = current_version + 1;
+            Self::truncate_ledger_db_single_batch(
+                &ledger_db,
+                &event_store,
+                &transaction_store,
+                start_version,
+                end_version,
+            )?;
+            current_version = start_version - 1;
+        }
+        assert!(current_version == target_version);
+        Ok(())
+    }
+
+    fn num_frozen_nodes_in_accumulator(num_leaves: u64) -> u64 {
+        2 * num_leaves - num_leaves.count_ones() as u64
+    }
+
+    fn truncate_transaction_accumulator(
+        ledger_db: &DB,
+        start_version: Version,
+        end_version: Version,
+        batch: &SchemaBatch,
+    ) -> Result<()> {
+        let num_frozen_nodes = Self::num_frozen_nodes_in_accumulator(end_version);
+        let mut iter = ledger_db.rev_iter::<TransactionAccumulatorSchema>(ReadOptions::default())?;
+        iter.seek_to_last();
```

Comment:
nit: can `.peek()` and you don't need to seek again

Reply:
but probably cleaner to seek to the start and use a for loop to assert forward in the first place

Reply:
Done.